### PR TITLE
release: build riscv64 binary, build static PIE if supported

### PR DESCRIPTION
**Pull request by kolyshkin**
This can be split into a few groups:
1. Dockerfile fixes.
2. Makefile fixes.
3. Enabling static PIE for arm64 and amd64.
4. Enabling riscv64.

See individual commits for details.

### CODE REVIEW

**Review by AkihiroSuda — APPROVED** [review]

**Comment by kolyshkin** [comment]
@cyphar @thaJeztah PTAL 🙏🏻

**Comment by AkihiroSuda** [comment]
> Enabling riscv64.

seccomp stuffs has to be updated too.
Opened PR https://github.com/opencontainers/runc/pull/3463

**Comment by thaJeztah** [comment]
> Opened PR https://github.com/opencontainers/runc/pull/3463

Slightly confused now; can these be merged independently, or should they be merged in a specific order? 😅

**Comment by AkihiroSuda** [comment]
> > Opened PR #3463
> 
> Slightly confused now; can these be merged independently, or should they be merged in a specific order? 😅

In arbitrary order, but merging one will need rebasing another one.

**Comment by kolyshkin** [comment]
Let me try to pick your changes @AkihiroSuda

**Review by AkihiroSuda — DISMISSED** [review]

**Comment by AkihiroSuda** [comment]
Needs rebase

**Comment by kolyshkin** [comment]
Rebased; PTAL @AkihiroSuda @thaJeztah @mrunalp

**Review by AkihiroSuda — APPROVED** [review]

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
This seems needed as well

```diff
diff --git a/libcontainer/seccomp/patchbpf/enosys_linux.go b/libcontainer/seccomp/patchbpf/enosys_linux.go
index dfb8a0a8..d1ee0c2b 100644
--- a/libcontainer/seccomp/patchbpf/enosys_linux.go
+++ b/libcontainer/seccomp/patchbpf/enosys_linux.go
@@ -67,6 +67,7 @@ const uint32_t C_AUDIT_ARCH_PPC64        = AUDIT_ARCH_PPC64;
 const uint32_t C_AUDIT_ARCH_PPC64LE      = AUDIT_ARCH_PPC64LE;
 const uint32_t C_AUDIT_ARCH_S390         = AUDIT_ARCH_S390;
 const uint32_t C_AUDIT_ARCH_S390X        = AUDIT_ARCH_S390X;
+const uint32_t C_AUDIT_ARCH_RISCV64      = AUDIT_ARCH_RISCV64;
 */
 import "C"
 
@@ -197,6 +198,8 @@ func archToNative(arch libseccomp.ScmpArch) (nativeArch, error) {
                return nativeArch(C.C_AUDIT_ARCH_S390), nil
        case libseccomp.ArchS390X:
                return nativeArch(C.C_AUDIT_ARCH_S390X), nil
+       case libseccomp.ArchRISCV64:
+               return nativeArch(C.C_AUDIT_ARCH_RISCV64), nil
        default:
                return invalidArch, fmt.Errorf("unknown architecture: %v", arch)
        }
```

**Comment by AkihiroSuda** [comment]
@kolyshkin Could you update the PR? ↑

**Comment by kolyshkin** [comment]
> @kolyshkin Could you update the PR? ↑

Sorry I missed it earlier. Updated.

**Review by AkihiroSuda — DISMISSED** [review]

**Comment by kolyshkin** [comment]
> ```diff
> +const uint32_t C_AUDIT_ARCH_RISCV64      = AUDIT_ARCH_RISCV64;
> ```

Alas this breaks the compilation on centos 7; I have added a workaround.

**Review by AkihiroSuda — DISMISSED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM

File: Makefile [hunk 1]
```diff
@@ -10,23 +10,51 @@ GIT_BRANCH_CLEAN := $(shell echo $(GIT_BRANCH) | sed -e "s/[^[:alnum:]]/-/g")
...
+GOARCH := $(shell $(GO) env GOARCH)
 
-ifeq ($(shell $(GO) env GOOS),linux)
-	ifeq (,$(filter $(shell $(GO) env GOARCH),mips mipsle mips64 mips64le ppc64))
-		ifeq (,$(findstring -race,$(EXTRA_FLAGS)))
-			GO_BUILDMODE := "-buildmode=pie"
-		endif
+GO_BUILDMODE :=
+# Enable dynamic PIE executables on supported platforms.
+ifneq (,$(filter $(GOARCH),386 amd64 arm arm64 ppc64le riscv64 s390x))
```
Comment (crazy-max) [inline Makefile:22]: afaik `-buildmode=pie` is not supported only on windows arm64 and linux mips*, ppc64be. should work on riscv64. see https://github.com/golang/go/blob/4aa1efed4853ea067d665a952eee77c52faac774/src/cmd/internal/sys/supported.go#L125-L131

Comment (kolyshkin) [reply Makefile:22]: See commit description at https://github.com/opencontainers/runc/pull/3446/commits/ab5c60d02fcd9a4b7e870bcd631c3673e5b7dd2a

Comment (kolyshkin) [reply Makefile:22]: @crazy-max PTAL ^^^

**Review by AkihiroSuda — APPROVED** [review]

**Comment by brandond** [comment]
Is there any particular reason that this hasn't been made available in a 1.x release yet? Trying to get k3s running on riscv64 and this looks like one of the last blockers.

**Comment by AkihiroSuda** [comment]
> release

This will be in v1.2.
I think we can have a RC soon.

**Comment by kolyshkin** [comment]
> Is there any particular reason that this hasn't been made available in a 1.x release yet? Trying to get k3s running on riscv64 and this looks like one of the last blockers.

This is a new feature, and we don't want to backport new features into an old release unless it's absolutely necessary.

If you need a binary, you can get one from any recent PR "Actions" tab. In there, click on "validate" in the left column, scroll all the way down to see "Artifacts", and in the zip will you will find a binary.

**Pull request merged**